docs/proposals/202101_endpoint_discovery.md

```diff
@@ -42,20 +44,25 @@ info:
     blah: ...
 ```
 
-Right now this info API would be added to existing components Ruler, Store, Sidecar, Compact, Downsample, Receive. But this would be added in the future for Targets, Exemplers, etc.
+Right now this info service would be added to existing components Ruler, Store, Sidecar, Compact, Downsample, Receive.
 
-The upgrade path would be possible in this case, as it's strictly additive for endpoints, and there will be a migration path to fallback to current discovery methods which could be removed in the future.
+The upgrade path would be possible in this case, as it's strictly additive for endpoints, and there will be a migration path to fall back to current discovery methods which could be removed in the future.
 
-#### Alternatives
+### Alternatives
 
-Split Info method out of the Store API into a separate gRPC service called `info`. Each component must implement and expose this. The discovery would happen via this info gRPC service, which each component would need to explicitly state which API it supports. Besides that information, we would also need to include here various info, e.g. LabelSets, Max/Min time, etc. The drawback for this solution is harder upgrades as this would be more complex to implement as we would need to fall back to older APIs.
+The plan would be for the initial discovery of gRPC services available to happen via gRPC reflection. To get further information, we would be adding an Info method to other services for example, Targets, Exemplars, etc. This would contain similar info or metadata information as we get right now from the Store Info method. We can always add more data to it in the future as needed.
```

Comment:
```suggestion
The plan would be for the initial discovery of gRPC services available to happen via gRPC reflection. To get further information, we would be adding an Info method to other services, for example, Targets, Exemplars, etc. This would contain similar info or metadata information as we get right now from the Store Info method. We can always add more data to it in the future as needed.
```

Reply:
So, will it happen via the Info service or gRPC reflection? I'm confused about this sentence :smile: Maybe I am missing/misunderstanding something. I thought with this proposal we will only use the Info service and we are going to try avoiding reflection

Reply:
Yes. We are only planning to use the Info service.
During discussions, It was also proposed that we can add Info methods (similar to the Info method we currently have in store API) to all other services (like rules, targets...) instead of creating a new Info service. Discovery of these methods may be done using gRPC reflection ( still depends on implementation 😅 ). That's why I added this as an alternative solution. 

I am also not sure about should we mention this solution as alternative solution? WDYT :)

Reply:
@GiedriusS we did a few changes in the proposal to make this clear. The TLDR is, we are going to use reflection to discover what services (StoreAPI, TargetsAPI, etc.) a particular endpoint is exposing. The alternative is to expose this information as part of the `Info` service itself. We are leaning towards reflection because in this case if some component adds support for a new service (ex. Store GW adds support for Exemplars) it will be automatically discovered by the Querier without extra work in code.

I think it's worth to explain why we are using reflection in the proposal as well.